### PR TITLE
Fix startAfter/endBefore for orderByKeys queries

### PR DESCRIPTION
**Pull request by IanWyszynski**
We should not be passing a fallback key to startAfter if we're querying on a key index. The indexValue is the key in this case.

### CODE REVIEW

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 1]
```diff
@@ -1390,6 +1390,50 @@ - (void)testStartAtPriorityAndEndAtPriorityWork {
...
 
+- (void)testStartAfterWithOrderByKey {
+  FIRDatabaseReference* ref = [FTestHelpers getRandomNode];
+  FIRDatabaseReference* childOne = [ref childByAutoId];
+  FIRDatabaseReference* childTwo = [ref childByAutoId];
+
+  [childOne setValue:@1];
+  [childTwo setValue:@2];
+
+  __block bool done = NO;
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:1401]: s/bool/BOOL

File: FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m [hunk 2]
```diff
@@ -1390,6 +1390,50 @@ - (void)testStartAtPriorityAndEndAtPriorityWork {
...
+
+- (void)testEndBeforeWithOrderByKey {
+  FIRDatabaseReference* ref = [FTestHelpers getRandomNode];
+  FIRDatabaseReference* childOne = [ref childByAutoId];
+  FIRDatabaseReference* childTwo = [ref childByAutoId];
+
+  [childOne setValue:@1];
+  [childTwo setValue:@2];
+
+  __block bool done = NO;
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m:1423]: s/bool/BOOL

File: FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m [hunk 3]
```diff
@@ -202,20 +202,26 @@ - (FIRDatabaseQuery *)queryStartingAfterValue:(id)startAfterValue {
...
+                initWithName:INVALID_QUERY_PARAM_ERROR
+                      reason:
+                          @"You must use queryStartingAfterValue: instead of "
+                          @"queryStartingAfterValue:childKey: when using "
+                          @"queryOrderedByKey:"
+                    userInfo:nil];
+        }
+        if ([startAfterValue isKindOfClass:[NSString class]]) {
+            startAfterValue = [FNextPushId successor:startAfterValue];
+        }
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m:217]: What happens when this is not true? Looks like we just call `queryStartingAtInternal` with the original value, which doesn't seem to do any validation.

Comment (IanWyszynski) [reply FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m:217]: This case is covered by `validateQueryEndpointsForParams`:

https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m#L106-L107

This is called in `queryingStartingAtInternal`.

File: FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m [hunk 4]
```diff
@@ -95,32 +95,36 @@ - (void)validateQueryEndpointsForParams:(FQueryParams *)params {
...
             }
             if (![params.indexStartValue.val isKindOfClass:[NSString class]]) {
                 [NSException
                      raise:INVALID_QUERY_PARAM_ERROR
-                    format:
-                        @"Can't use queryStartingAtValue: with other types "
-                        @"than string in combination with queryOrderedByKey"];
+                    format:@"Can't use queryStartingAtValue: or "
+                           @"queryStartingAfterValue: "
+                           @"with other types than string in combination with "
```
Comment (schmidt-sebastian) [inline FirebaseDatabase/Sources/Api/FIRDatabaseQuery.m:108]: Might be easier to understand if it said "with non-string types when used with queryOrderedByKey:"

**Review by schmidt-sebastian — APPROVED** [review]
Thanks for updating the error message.

Can you add a Changelog entry before submitting?

File: FirebaseDatabase/Sources/Core/FRepo.m [hunk 5]
```diff
@@ -518,10 +518,8 @@ - (void)getData:(FIRDatabaseQuery *)query
         (void (^_Nonnull)(NSError *__nullable error,
                           FIRDataSnapshot *__nullable snapshot))block {
     FQuerySpec *querySpec = [query querySpec];
-    id<FNode> node =
-        [self.serverSyncTree calcCompleteEventCacheAtPath:querySpec.path
-                                          excludeWriteIds:@[]];
-    if (![node isEmpty]) {
+    id<FNode> node = [self.serverSyncTree getServerValue:[query querySpec]];
+    if (node != nil && ![node isEmpty]) {
```
Comment (IanWyszynski) [inline FirebaseDatabase/Sources/Core/FRepo.m:522]: @schmidt-sebastian I'm stuck here. I can't figure out why `[node isEmpty]` is needed to pass the two tests:

https://github.com/firebase/firebase-ios-sdk/pull/7403/files#diff-3a8cb26bda8f6a6d6700b7c2ee8b1368a1c147ed7bb29743a0f6146c08d7b723R1437-R1495

The equivalent is not needed for the android changes.

**Review by schmidt-sebastian — COMMENTED** [review]
I still have do debug the [node isEmpty] issue, but it looks like the SDK doesn't read from cache when it should.

**Pull request merged**